### ISSUE TITLE
a few questions about the math

### ISSUE THREAD

**Issue by shenyueshi** [hongzimao/pensieve#4]
1) The training law of the Actor network (Eq.2) uses the gradient of the network times the reward difference A, to evolve the model of the neural network

Is the term "detla_theda log pi_theda (s_t, a_t)" the gradient of the network?

I am a little confused about what's in ac3.py: actor_gradients is pi_theda(s_t, a_t), then what does tf.log(tf.reduce_sum(tf.multiply(self.out, self.acts),...)) do? Which term is the function A?

In a3c.py, line 46 - 59

2) The value of A(s_t, a_t) is calculated by 
r_t + gamma V_pi_theda(s_t+1; theda_v) - V_pi_theda(st;theda_v): the expression between Eq.3 and 4

The first term r_t is the new reward based on the new action a. The second term is the future reward based on the new state s_t+1, but with the policy unchanged (no more new action)?

3) The training of the critic network is based on "empirically observed rewards", which is the value of r_t (based on s_t and a_t) in Eq.3

In the ac3.py line 144, where does td_target come from?

Suppose the "empirically observed rewards" is the ground truth, can supervised learning also work? What is the value of using reinforcement learning?

4) A sperate question about how sensitive the trained model to different parameters. The state s_t contains a number of data fields, which one do you think brings the greatest impact on improvement (e.g., knowing the sizes of m future chunks might be very useful)? Also, the number of past chunks and the number of future chunks should affect the complexity and the accuracy of the neural network, what do you think is the sweet spot?

A related question is that you use the "Envivo-Dash3" video set, do you think a different transcoder stack with a different bitrate ladder, segment duration, even the fluctuation of bitrate (within one bitrate) may make a trained model perform less well?

**Comment by hongzimao** [comment]
1. Yes, detla_theda log pi_theda (s_t, a_t) advantage(s_t, a_t) is the gradient for actor network. The term for advantage function A is `self.act_grad_weights`, which is passed through `td_batch` from  https://github.com/hongzimao/pensieve/blob/master/sim/a3c.py#L240. 

2. The policy is unchanged, because it is evaluated on the same actor network. However, "no more action" is incorrect--V_pi is the **value function**, which estimates the cumulated reward of the remaining episode. 

3. `td_target` is obtained from the total empirical rewards in the actual experience. In the code, it's passed from https://github.com/hongzimao/pensieve/blob/master/sim/a3c.py#L243. Indeed, the critic is trained kind of using "supervised learning" (ground truth obtained via what is known as temporal difference). But keep in mind that this is for the critic, which aids the actor to estimate how good the actions are. Also, even here the "ground truth" is not readily available in the environment, it's a consequence of the interaction. Thus, it is not the same meaning as "labels" in supervised learning. Overall, the whole framework is reinforcement learning.

4. I would say (intuitively) the _current buffer size_ is the most important. But we didn't extensively analyze this. You can try to eliminate a specific term, play with the system, and see how large the performance degradation would be. We evaluate the impact of number of past chunks in Figure 14 and Section 5.4 in the paper. Ideally, we would use LSTM/RNN to process time-series sequences, instead of a fixed size sequence. For your last point regarding varying video conditions, we actually extend and evaluate the learning algorithm across multiple videos. Like you said, those videos have very different properties (e.g., number of bitrates, bitrate levels, video length, etc.). In our paper, section 4.3 describes the technique we used for multi-video extension. Figure 13 and section 5.3 shows the evaluation results. The code is open sourced in https://github.com/hongzimao/pensieve/tree/master/multi_video_sim.
